### PR TITLE
Raise an exception message with JSON is invalid.

### PR DESCRIPTION
**Pull request by aaron-collier**
This raises an error with the HTTP response is a 200 but the JSON received is invalid. 

Addressing issue #241

### CODE REVIEW

File: app/services/iiif_metadata_service.rb [hunk 1]
```diff
@@ -53,5 +53,7 @@ def json
                   data['@id'] = @canonical_url
                 end
               end
+  rescue JSON::ParserError
+    raise "There was a problem, #{@url} returned invalid JSON."
```
Comment (jcoyne) [inline app/services/iiif_metadata_service.rb]: Can you provide the response in the message too? Also, how about making a custom error?  How about `Stacks::UnexpectedMetadataResponse`?

Comment (aaron-collier) [reply app/services/iiif_metadata_service.rb]: For sure.

File: app/services/stacks/unexpected_metadata_response.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+# A custom exceptions class to provide specific Stacks::Exceptions
+module Stacks
+  # An UnexpectedMetadataResponse to raise when HTTP returns bad JSON
+  class UnexpectedMetadataResponse < StandardError
+    def initialize(url = nil, message = nil)
```
Comment (jcoyne) [inline app/services/stacks/unexpected_metadata_response.rb]: When you call this you are passing a `ParserError` to the argument you are  calling `message`

File: spec/services/iiif_metadata_service_spec.rb [hunk 3]
```diff
@@ -8,42 +8,64 @@
...
+    before do
+      allow(HTTP).to receive(:get)
+        .with('https://sul-imageserver-uat.stanford.edu/cantaloupe/iiif/2/nr%2F349%2Fct%2F7889%2Fnr349ct7889_00_0001.jp2/info.json')
+        .and_return(bad_response)
+    end
+
+    describe "#fetch" do
+      subject { service.fetch(256) }
+      it "raises Stacks::UnexpectedMetadataResponse" do
+        expect { subject }.to raise_error { |error|
```
Comment (jcoyne) [inline spec/services/iiif_metadata_service_spec.rb]: change to `expect { subject }.to raise_error Stacks::UnexpectedMetadataResponse`

File: app/services/iiif_metadata_service.rb [hunk 4]
```diff
@@ -49,10 +49,13 @@ def retrieve
...
   def json
+    retrieved_json = retrieve
     @json ||= begin
-                JSON.parse(retrieve).tap do |data|
+                JSON.parse(retrieved_json).tap do |data|
                   data['@id'] = @canonical_url
                 end
               end
+  rescue JSON::ParserError => error
+    raise Stacks::UnexpectedMetadataResponseError.new(@url, error, retrieved_json)
```
Comment (cbeer) [inline app/services/iiif_metadata_service.rb]: I don't think we need to pass `error` as a parameter; it should be available in the exception as a `cause`, maybe?

**Pull request merged**